quinn-proto/src/token.rs

```diff
@@ -30,50 +32,98 @@ impl IncomingTokenState {
     }
 }
 
-pub(crate) struct RetryToken {
-    /// The destination connection ID set in the very first packet from the client
-    pub(crate) orig_dst_cid: ConnectionId,
-    /// The time at which this token was issued
-    pub(crate) issued: SystemTime,
+/// An address validation / retry token
+///
+/// The data in this struct is encoded and encrypted in the context of not only a handshake token
+/// key, but also a client socket address.
+pub(crate) struct Token {
+    /// Randomly generated value, which must be unique, and is visible to the client
+    pub(crate) rand: u128,
+    /// Content which is encrypted from the client
+    pub(crate) inner: RetryTokenInner,
 }
 
-impl RetryToken {
-    pub(crate) fn encode(
-        &self,
-        key: &dyn HandshakeTokenKey,
-        address: &SocketAddr,
-        retry_src_cid: &ConnectionId,
-    ) -> Vec<u8> {
-        let aead_key = key.aead_from_hkdf(retry_src_cid);
+impl Token {
```

Comment:
This commit still has a lot going on. Would it be feasible to split this into smaller parts? I guess I'm looking for one that sort of changes the public interface and another one that changes the internal logic, as much as possible.

Reply:
Yeah, it could help make it more gradual if I made one commit that just converts it from:

```rs
struct Token {
    pub(crate) orig_dst_cid: ConnectionId,
    pub(crate) issued: SystemTime,
}
```

To:

```rs
struct Token {
    pub(crate) inner: RetryTokenInner,
}

struct RetryTokenInner {
    pub(crate) orig_dst_cid: ConnectionId,
    pub(crate) issued: SystemTime,
}
```

And then a second commit that adds the `rand: u128` field to `Token` and change how it's encrypted. And there's various other variations on this I could do.

It's an ideological question about how much you want each commit to be fully sensible on its own, versus how much a commit may do slightly strange seeming things in the name of setting up for subsequent commits. Making `Token` into a struct that only has one field is kind of weird on its own, but it makes subsequent commits easier. I've worked with people who would've objected to such a thing.

A similar principle applies to your subsequent comment of "This was already renamed once before, can we get the name right in the earlier commit?" w.r.t. `RetryToken` -> `RetryTokenInner` -> `TokenInner`.

Reply:
I find "ideological" a bit loaded, but I agree there's a trade-off here. Personally, I value having smaller commits a lot since IME it makes reviewing so much easier and I think reviews are highly valuable (and more likely to catch issues if they're easy). At the same time, reverting within-PR changes is suboptimal because it increases the amount of change to review. To acommodate all that, I think we're usually willing to sustain some commits with stylistic (including naming) choices that only make sense in the context of the PR/grander purpose of the goal (and it can be useful to note that trade-off in the commit message).